go/vt/vtexplain/vtexplain_vtgate.go

```diff
@@ -0,0 +1,142 @@
+	vschemapb "github.com/youtube/vitess/go/vt/proto/vschema"
+	vtgatepb "github.com/youtube/vitess/go/vt/proto/vtgate"
+)
+
+var (
+	explainTopo    *ExplainTopo
+	vtgateExecutor *vtgate.Executor
+	healthCheck    *discovery.FakeHealthCheck
+
+	vtgateSession = &vtgatepb.Session{
+		TargetString: "@master",
+		Autocommit:   true,
+	}
+)
+
+func initVtgateExecutor(vSchemaStr string, opts *Options) error {
+	explainTopo = &ExplainTopo{NumShards: opts.NumShards}
+	healthCheck = discovery.NewFakeHealthCheck()
+
+	resolver := newFakeResolver(healthCheck, explainTopo, vtexplainCell)
+
+	err := buildTopology(vSchemaStr, opts.NumShards)
+	if err != nil {
+		return err
+	}
+
+	streamSize := 10
+	vtgateExecutor = vtgate.NewExecutor(context.Background(), explainTopo, vtexplainCell, "", resolver, opts.Normalize, streamSize)
+
+	return nil
+}
+
+func newFakeResolver(hc discovery.HealthCheck, serv topo.SrvTopoServer, cell string) *vtgate.Resolver {
+	gw := gateway.GetCreator()(hc, topo.Server{}, serv, cell, 3)
+	gw.WaitForTablets(context.Background(), []topodatapb.TabletType{topodatapb.TabletType_REPLICA})
+	tc := vtgate.NewTxConn(gw, vtgatepb.TransactionMode_MULTI)
+	sc := vtgate.NewScatterConn("", tc, gw)
+	return vtgate.NewResolver(serv, cell, sc)
+}
+
+func buildTopology(vschemaStr string, numShardsPerKeyspace int) error {
+	explainTopo.Lock.Lock()
+	defer explainTopo.Lock.Unlock()
+
+	explainTopo.Keyspaces = make(map[string]*vschemapb.Keyspace)
+	err := json.Unmarshal([]byte(vschemaStr), &explainTopo.Keyspaces)
+	if err != nil {
+		return err
+	}
+
+	explainTopo.TabletConns = make(map[string]*sandboxconn.SandboxConn)
+	for ks, vschema := range explainTopo.Keyspaces {
+		numShards := 1
+		if vschema.Sharded {
+			numShards = numShardsPerKeyspace
+		}
+		for i := 0; i < numShards; i++ {
+			kr, err := key.EvenShardsKeyRange(i, numShards)
+			if err != nil {
+				return err
+			}
+			shard := key.KeyRangeString(kr)
+			hostname := fmt.Sprintf("%s/%s", ks, shard)
+			log.Infof("registering test tablet %s for keyspace %s shard %s", hostname, ks, shard)
+			sc := healthCheck.AddTestTablet(vtexplainCell, hostname, 1, ks, shard, topodatapb.TabletType_MASTER, true, 1, nil)
+			explainTopo.TabletConns[hostname] = sc
+		}
+	}
+
+	return err
+}
+
+func vtgateExecute(sql string) ([]*engine.Plan, map[string][]*TabletQuery, error) {
+	_, err := vtgateExecutor.Execute(context.Background(), vtgateSession, sql, nil)
+	if err != nil {
+		return nil, nil, fmt.Errorf("vtgate Execute: %v", err)
+	}
+
+	// use the plan cache to get the set of plans used for this query, then
+	// clear afterwards for the next run
+	planCache := vtgateExecutor.Plans()
+	plans := make([]*engine.Plan, 0, 4)
+	for _, item := range planCache.Items() {
+		plans = append(plans, item.Value.(*engine.Plan))
+	}
+	planCache.Clear()
+
+	tabletQueries := make(map[string][]*TabletQuery)
+	for tablet, tc := range explainTopo.TabletConns {
+		if len(tc.Queries) == 0 {
```

Comment:
Yes it can in case where the given tablet never got any queries in the current plan. The way the system works is it creates fake tablets for all keyspaces / shards. So unless you have a scatter query it's likely some shards got no queries.

The check isn't strictly necessary, it just avoids putting empty entries in the `tabletQueries` map.